### ISSUE TITLE
Potential race condition in function sslize with OpenSSL

### ISSUE THREAD

**Issue by photools** [civetweb/civetweb#493]
I'm using CivetWeb for quite some time now. Windows, Visual Studio 2015, 32-bit, current version of civetweb.

## The Problem

Recently I've converted the application which hosts CivetWeb to the 64-bit platform. While doing stress tests I've encountered **frequent crashes in the sslize  function** in civetweb.c. But only when compiling for 64-bit, never had that with 32-bit.

## What I Already Did

I first made sure that I have the most recent version of OpenSSL (1.1.0.2) installed and than the correct DLLs are loaded by CW. I use the binaries provided by [Shining Light](https://slproweb.com/products/Win32OpenSSL.htmll) for many years. 32-bit and 64-bit.

My code compiles warning free (level 3) and the static code analysis of Visual Studio also comes up clean. (except for the few minor issues in CivertWeb I've reported recently). I've even run my application under App Verifier to check for memory overruns and other hard to find issues.

## Producing the Crash

My ´test suite includes a custom-written JavaScript app which simulates a user performing normal operations. This app can be run multiple times to simulate multiple users and to put stress on my web service, and CivetWeb.

And when I create enough stress (say 5 to 10 users) CivetWeb _reproducibly_ crashes with a GPF inside (!) ssleay32.dll, four levels deep in the call stack:

    for (i = 16; i <= 1024; i *= 2)  {
    ret = func(conn->ssl);  // <== Here it crashes
    ...
I noticed that this always happens when 
* the loop index is 16 or 32 and 
* when there are **multiple calls to sslize on the call stack**. Usually 3 or more.

The conn variable and ssl variable look OK in the debugger, if one can tell in optimized (release mode) code.

It takes about 2 to 3 minutes for my test suite to produce the crash.

## The Solution (?)

OpenSSL has a history for multi-threaded issues, even when (as CivetWeb does!) the MT callbacks are provided. So. for testing, I've added a CRITICAL_SECTION (Windows) to the sslize function, protecting the calls into OpenSSL in that function. **And the problem is gone. No crashes anymore.**

But probably this is a recipe for deadlocks, caused by side effects...

**Comment by bel2125** [comment]
I am not sure what should be done here - if it's clearly an OpenSSL bug, it must be handled there.

The function "func" returns some "SSL_ERROR_WANT_*" error for several times before - do you know which WANT_* exactly? The crash in in the "sslize" call from "worker_thread_run" with "func" == "SSL_accept", right?

Is this issue gone with a different version of OpenSSL? E.g. 1.0.something?
Adding or not adding a lock may have different effects on different versions of OpenSSL.

**Comment by photools** [comment]
The hack with protecting the sslize function with a critical section caused dead-locks, as I anticipated. Such hacks never work. The CivetWeb code looks sound and from what I can tell it should just work.

I've studied the OpenSSL web site and release history and learned that they did work on the multi-threading support and *internalized' shared resource protection in the 1.1 version. So I decided to bite the bullet and migrated my code to OpenSSL 1.1 yesterday - and enabled 1.1 support in CivetWeb as well.

**And this works just fine.** 

No more issues with CivertWeb, crashing in OpenSLL calls even when I run it under high stress for elongated periods. This is no as stable as CivetWeb without SLL. Awesome.

I can even use OpenSSL 1.0 and 1.1 simultaneously  (I use a 3rd party email component which requires OpenSSL 1.0).
